cpp/include/cuspatial/experimental/detail/linestring_distance.cuh

```diff
@@ -0,0 +1,284 @@
+
+  auto const& A = thrust::raw_reference_cast(linestring1_points_begin[p1_idx]);
+  auto const& B = thrust::raw_reference_cast(linestring1_points_begin[p1_idx + 1]);
+
+  auto min_squared_distance = std::numeric_limits<T>::max();
+  for (auto p2_idx = ls2_start; p2_idx < ls2_end; p2_idx++) {
+    auto const& C        = thrust::raw_reference_cast(linestring2_points_begin[p2_idx]);
+    auto const& D        = thrust::raw_reference_cast(linestring2_points_begin[p2_idx + 1]);
+    min_squared_distance = std::min(min_squared_distance, squared_segment_distance(A, B, C, D));
+  }
+
+  atomicMin(&thrust::raw_reference_cast(*(distances + linestring_idx)),
+            static_cast<T>(std::sqrt(min_squared_distance)));
+}
+
+}  // namespace detail
+
+template <class Cart2dItA, class Cart2dItB, class OffsetIterator, class OutputIt>
+void pairwise_linestring_distance(OffsetIterator linestring1_offsets_first,
+                                  OffsetIterator linestring1_offsets_last,
+                                  Cart2dItA linestring1_points_first,
+                                  Cart2dItA linestring1_points_last,
+                                  OffsetIterator linestring2_offsets_first,
+                                  Cart2dItB linestring2_points_first,
+                                  Cart2dItB linestring2_points_last,
+                                  OutputIt distances_first,
+                                  rmm::cuda_stream_view stream)
+{
+  using T = typename std::iterator_traits<Cart2dItA>::value_type::value_type;
+
+  static_assert(
+    detail::is_floating_point<T,
+                              typename std::iterator_traits<Cart2dItB>::value_type::value_type,
+                              typename std::iterator_traits<OutputIt>::value_type>(),
+    "Inputs and output must be floating point types.");
+
+  static_assert(detail::is_same<T,
+                                typename std::iterator_traits<Cart2dItB>::value_type::value_type,
+                                typename std::iterator_traits<OutputIt>::value_type>(),
+                "Inputs and output must be the same types.");
+
+  static_assert(detail::is_same<cartesian_2d<T>,
+                                typename std::iterator_traits<Cart2dItA>::value_type,
+                                typename std::iterator_traits<Cart2dItB>::value_type>(),
+                "Inputs must be cuspatial::cartesian_2d");
+
+  auto const num_string_pairs =
+    thrust::distance(linestring1_offsets_first, linestring1_offsets_last);
+  auto const num_linestring1_points =
+    thrust::distance(linestring1_points_first, linestring1_points_last);
+  auto const num_linestring2_points =
+    thrust::distance(linestring2_points_first, linestring2_points_last);
+
+  thrust::fill(rmm::exec_policy(stream),
+               distances_first,
+               distances_first + num_string_pairs,
+               std::numeric_limits<T>::max());
+
+  std::size_t constexpr threads_per_block = 64;
```

Comment:
We use `std::size_t` here while using `::size_t` in the kernel. We should probably be consistent.

Reply:
kernel should use the former.